programs/mango-v4/src/state/bank.rs

```diff
@@ -693,16 +733,40 @@ impl Bank {
         allow_dusting: bool,
         now_ts: u64,
     ) -> Result<WithdrawResult> {
-        let opening_indexed_position = position.indexed_position;
-        let res = self.withdraw_internal(
-            position,
-            native_amount,
-            with_loan_origination_fee,
-            allow_dusting,
-            now_ts,
-        );
-        self.update_cumulative_interest(position, opening_indexed_position);
-        res
+        if position.allow_lending() {
+            assert!(position.unlendable_deposits == 0);
+            let opening_indexed_position = position.indexed_position;
+            let res = self.withdraw_internal(
+                position,
+                native_amount,
+                with_loan_origination_fee,
+                allow_dusting,
+                now_ts,
+            );
+            self.update_cumulative_interest(position, opening_indexed_position);
+            res
+        } else {
+            assert!(position.indexed_position.is_zero());
+
+            // Note: this rounds up native_amount, withdrawing a fraction of a token more
+            // than desired and dusting the difference!
+            let withdraw_amount = native_amount.ceil();
+            self.dust += withdraw_amount - native_amount;
+
+            let withdraw_amount_u64 = withdraw_amount.to_num::<u64>();
+            require_gte!(
+                position.unlendable_deposits,
+                withdraw_amount_u64,
+                MangoError::UnlendableTokenPositionCannotBeNegative
+            );
+            position.unlendable_deposits -= withdraw_amount_u64;
+            self.unlendable_deposits -= withdraw_amount_u64;
```

Comment:
Should/could we also add a require_gte on the bank unlendable deposit vs withdrawn amount ?